File: 11-container-with-most-water/container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
// Two-Pointer technique
// Time: O(n)
// Space: O(1)
var maxArea = function(height) {
    let s = 0, e = height.length-1
    let max = 0
    while (s < e) {

        max = Math.max(max, Math.min(height[s], height[e]) * (e - s))
        if (height[s] > height[e]) {
            e--
        } else {
            s++
        }
    }
    return max
};

File: 11-container-with-most-water/container-with-most-water.swift
class Solution {
    func maxArea(_ height: [Int]) -> Int {
        var area: Int = 0
        var lo: Int = 0, hi = height.count - 1

        while lo < hi {
            var h = min(height[lo], height[hi]) 
            var w: Int = hi - lo
            area = max(area, h * w)
            height[lo] <= height[hi] ? (lo += 1) : (hi -= 1)
        }
        return area
    }
}